rpc/providers/web-socket-proxy.test.ts

```diff
@@ -0,0 +1,114 @@
+import { deferred } from "../../deps/std/async.ts";
+import { polkadot } from "../../known/mod.ts";
+import { getRandomPort } from "../../test-util/mod.ts";
+import { WebSocketProxy } from "./web-socket-proxy.ts";
+
+Deno.test({
+  name: "WebSocketProxy",
+  async fn(t) {
+    await t.step({
+      name: "reconnect on client-side WebSocket close",
+      async fn() {
+        let firstConnection: WebSocket;
+
+        const webSocketFactory = () => {
+          const ws = new WebSocket(polkadot.discoveryValue);
+
+          if (!firstConnection) {
+            firstConnection = ws;
+          }
+
+          return ws;
+        };
+
+        const webSocketProxy = new WebSocketProxy({
+          webSocketFactory,
+          reconnect: {
+            delay: 0,
+          },
+        });
+
+        const isOpen = deferred();
+        webSocketProxy.onopen = () => isOpen.resolve();
+
+        await isOpen;
+
+        const isReopen = deferred();
+        webSocketProxy.onopen = () => isReopen.resolve();
+
+        const isClosed = deferred();
+        webSocketProxy.onclose = () => isClosed.resolve();
+
+        firstConnection!.close();
```

Comment:
It might be better to change this to a `spy`.
Any thoughts on 
```
import sinon from "https://cdn.skypack.dev/sinon";
```

Reply:
Sounds reasonable :) which functionality do you want to stub?

Reply:
I would use a spy rather than a stub.
So, I would be able to spy on the WebSocketFactory return value and invoke `.close`.
Also spying on the factory would let me assert that it was invoked 2 times due to a disconnection.